tests/devices/qubit/test_measure.py

```diff
@@ -0,0 +1,148 @@
+        """Test that `Sum` expectation values are correct when tensor contraction
+        is used for computation."""
+        summands = (qml.prod(qml.PauliY(i), qml.PauliZ(i + 1)) for i in range(7))
+        obs = qml.sum(*summands)
+        ops = [qml.RX(0.123, wires=i) for i in range(8)]
+        meas = [qml.expval(obs)]
+        qs = qml.tape.QuantumScript(ops, meas)
+
+        res = simulate(qs)
+        expected = 7 * (-np.sin(0.123) * np.cos(0.123))
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize(
+        "obs, expected",
+        [
+            (qml.sum(qml.PauliY(0), qml.PauliZ(0)), -np.sin(0.123) + np.cos(0.123)),
+            (
+                qml.sum(*(qml.PauliZ(i) for i in range(8))),
+                sum(np.sin(i * np.pi / 2 + 0.123) for i in range(8)),
+            ),
+        ],
+    )
+    def test_sum_expval_eigs(self, obs, expected):
+        """Test that `Sum` expectation values are correct when eigenvalues are used
+        for computation."""
+        ops = [qml.RX(i * np.pi / 2 + 0.123, wires=i) for i in range(8)]
+        meas = [qml.expval(obs)]
+        qs = qml.tape.QuantumScript(ops, meas)
+
+        res = simulate(qs)
+        assert np.allclose(res, expected)
+
+    @pytest.mark.parametrize(
+        "meas, expected",
+        [
+            (qml.state(), "state_measurement_process"),
+            (qml.probs(wires=[0, 1]), "state_measurement_process"),
+            (qml.density_matrix(wires=[0]), "state_measurement_process"),
+            (qml.expval(qml.PauliZ(0)), "state_diagonalizing_gates"),
+            (
+                qml.var(qml.s_prod(2.5, qml.prod(qml.PauliZ(0), qml.PauliX(1)))),
+                "state_diagonalizing_gates",
+            ),
+            (
+                qml.expval(qml.Hamiltonian([-0.5, 2], [qml.PauliY(0), qml.PauliZ(0)])),
+                "state_hamiltonian_expval",
+            ),
+            (
+                qml.expval(
+                    qml.SparseHamiltonian(
+                        csr_matrix(-0.5 * qml.PauliY(0).matrix() + 2 * qml.PauliZ(0).matrix()),
+                        wires=[0],
+                    )
+                ),
+                "state_hamiltonian_expval",
+            ),
+            (
+                qml.expval(
+                    qml.sum(*(qml.prod(qml.PauliY(i), qml.PauliZ(i + 1)) for i in range(7)))
+                ),
+                "state_hamiltonian_expval",
+            ),
+            (qml.expval(qml.sum(qml.PauliY(0), qml.PauliZ(0))), "state_diagonalizing_gates"),
+            (qml.expval(qml.sum(*(qml.PauliZ(i) for i in range(8)))), "state_diagonalizing_gates"),
+        ],
+    )
+    def test_correct_expval_measurement_used(self, meas, expected, mocker):
+        """Test that the correct measurement function is used for a given observable."""
+        qs = qml.tape.QuantumScript(measurements=[meas])
+        spy = mocker.spy(qml.devices.qubit.measure, expected)
```

Comment:
We no longer need to use `mocker.spy` to test the code. Instead we can unit test the `get_measurement_function` and make sure that function dispatches us to the right method.

If we know that `get_measurement_function` dispatches us to the correct internal function, and that the internal function gives us correct results, and the integrated `measure` function gives us correct results, I think we are good to go. No mocker needed.